internal/provider/resource_vpc_firewall_rules_v0.go

```diff
@@ -0,0 +1,724 @@
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/oxidecomputer/oxide.go/oxide"
+)
+
+// Schemas and model definitions for upgrading the oxide_vpc_firewall_rules
+// resource from version 0.
+//
+// Version 0.13.0 of the provider had breaking changes to the schema without
+// a version bump, so there are two different schemas versioned as 0. The
+// original schema (present in provider versions <0.13.0) is referred to here
+// as v0.0, and the schema with breaking changes (present in provider version
+// =0.13.0) is referred to as v0.1
+//
+// Schemas and structs are mostly copied from their original sources.
+// https://github.com/oxidecomputer/terraform-provider-oxide/blob/v0.12.0/internal/provider/resource_vpc_firewall_rules.go
+// https://github.com/oxidecomputer/terraform-provider-oxide/blob/v0.13.0/internal/provider/resource_vpc_firewall_rules.go
+
+// stateUpgraderV0 is a StateUpgrader function to upgrades an
+// oxide_vpc_firewall_rules resource from schema version 0 to latest.
+//
+// It must be able to handle both versions of schema v0. States in schema v0.0
+// are first upgraded to v0.1, and then the v0.1 state is upgraded to the
+// latest schema version. This way future upgrades only need to handle
+// upgrades from v0.1.
+func (r *vpcFirewallRulesResource) stateUpgraderV0(ctx context.Context, req resource.UpgradeStateRequest, resp *resource.UpgradeStateResponse) {
+	var modelV01 vpcFirewallRulesResourceModelV01
+
+	// Check if we need to upgrade from v0.0 to v0.1 first.
+	schemaV00 := r.schemaV00(ctx)
+	rawStateValue, err := req.RawState.Unmarshal(schemaV00.Type().TerraformType(ctx))
+	if err == nil {
+		stateV00 := &tfsdk.State{
+			Raw:    rawStateValue,
+			Schema: schemaV00,
+		}
+
+		var modelV00 vpcFirewallRulesResourceModelV00
+		resp.Diagnostics.Append(stateV00.Get(ctx, &modelV00)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+
+		// Upgrade from schema v0.0 to v0.1.
+		var diags diag.Diagnostics
+		modelV01, diags = modelV00.upgrade(ctx)
+		resp.Diagnostics.Append(diags...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	} else {
+		tflog.Info(ctx, "failed to unmarshal state using schema v0.0, trying v0.1", map[string]any{"err": err})
+
+		// Unmarshalling to schema v0.0 failed, try with schema v0.1.
+		schemaV01 := r.schemaV01(ctx)
+		rawStateValue, err := req.RawState.Unmarshal(schemaV01.Type().TerraformType(ctx))
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Unable to Upgraded State From Version 0",
+				fmt.Sprintf("failed to convert state to schema version 0.1: %v", err),
+			)
+			return
+		}
+
+		stateV01 := &tfsdk.State{
+			Raw:    rawStateValue,
+			Schema: schemaV01,
+		}
+		resp.Diagnostics.Append(stateV01.Get(ctx, &modelV01)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+
+	// Upgrade from schema v0.1 to latest.
+	newState := modelV01.upgrade()
+	resp.Diagnostics.Append(resp.State.Set(ctx, newState)...)
```

Comment:
Just making sure I'm following this correctly. This v0.1 upgrade is really a no-op outside of type changes, correct? I see that v0.0 to v0.1 accounts for the new `type` field in the JSON but I do not see us doing anything special with v0.1 states. Just wanted to confirm.